src/org/rascalmpl/library/lang/html/IO.java

```diff
@@ -113,23 +150,59 @@ else if (node instanceof DataNode) {
 
         IListWriter w = factory.listWriter();
         for (Node n : elem.childNodes()) {
-            w.append(toConstructorTree(n));
+            if (!(n instanceof Comment)) {
+                w.append(toConstructorTree(n, file, includeEndTags));
+            }
         }
-
-        if (cons.getArity() > 0) {
-            return factory.constructor(cons, new IValue[] { w.done() }, kws);
+                
+        IConstructor result = cons.getArity() > 0
+            ? factory.constructor(cons, new IValue[] { w.done() }, kws)
+            : factory.constructor(cons, new IValue[0], kws);
+        
+        if (file != null) {
+            ISourceLocation src = nodeToLoc(elem.sourceRange(), elem.endSourceRange(), file, includeEndTags);
+            return result.asWithKeywordParameters().setParameter("location", src);
+        } else {
+            return result;
         }
-        else {
-            return factory.constructor(cons, new IValue[0], kws);
+    }
+
+    private ISourceLocation nodeToLoc(Range r, Range e, ISourceLocation file, boolean includeEndTags) {
+        if (!r.isTracked()) {
+            return file;
         }
+
+        return includeEndTags && e.isTracked()
+            ? factory.sourceLocation(file,
```

Comment:
there appear to be no difference between these two branches?

Reply:
the variable names are too short too notice the difference. I'll refactor this.